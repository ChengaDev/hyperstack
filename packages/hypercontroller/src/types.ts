import type { Request as XRequest } from 'express'
import { NextFunction, Response } from 'express'

interface Request extends XRequest {
  id: any
  user?: any
  state?: any
  logger?: any
}

export { Request, Response, NextFunction }

export interface RequestIdOpts {
  mode: 'none' | 'passthrough' | 'passthrough-or-generate' | 'generate'
}

export interface ServerOptions {
  logging?: { logger?: boolean | any; middleware?: boolean | any } // true | {logger, opts}
  staticFolder?: string
  indexCatchAll?: boolean
  forceHttps?: boolean
  cookieSecret?: string
  sendValidationErrors?: boolean
  gzip?: boolean | any
  helmet?: boolean | any
  json?: boolean | any
  requestId?: boolean | RequestIdOpts
  urlencoded?: boolean | any
  beforeControllers?: any
  afterControllers?: any
  beforeMiddleware?: any
  afterMiddleware?: any
}

export interface CreateServerOptions {
  controllers: any[]
  opts?: ServerOptions
}
