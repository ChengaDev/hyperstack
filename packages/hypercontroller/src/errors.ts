class ValidationError extends Error {
  readonly isValidationError = true

  constructor(public errors?: any) {
    super()
    Object.setPrototypeOf(this, ValidationError.prototype)
  }
}

export { ValidationError }
