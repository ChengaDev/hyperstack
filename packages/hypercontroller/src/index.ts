export * from './http/responses' // as Responses
export * from './validation'
export * from './errors'
export * from './types'
export * from './api'
export * from './jwt'
export * from './routes'
