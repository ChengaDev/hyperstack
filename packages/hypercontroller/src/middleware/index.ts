import cookieParserMiddleware from 'cookie-parser'
import asyncWrapMiddleware from 'async-express-mw'
import { static as staticMiddleware } from 'express'
import forceHttpsMiddleware from './force-https'

export {
  staticMiddleware,
  cookieParserMiddleware,
  asyncWrapMiddleware,
  forceHttpsMiddleware,
}
