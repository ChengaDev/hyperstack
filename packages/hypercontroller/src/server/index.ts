import { HyperServer, createServer } from './hyperserver'
import listRoutes from './list-routes'
import printRoutes from './print-routes'
import load from './load'
import listen from './listen'

export { HyperServer, createServer, listRoutes, load, listen, printRoutes }
