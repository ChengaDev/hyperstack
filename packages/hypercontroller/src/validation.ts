import Ajv from 'ajv'
import L from 'lodash'
import { Middleware } from '@hyperstackjs/hypernight'
import type { ZodRawShape, z } from 'zod'
import { ValidationError } from './errors'
import type { NextFunction, Request, Response } from './types'
// strong params, permit only what's in schema
const permits = <T extends ZodRawShape>(schema: z.ZodObject<T>) =>
  schema.parse.bind(schema)

// actionparams, permit and require all that's in schema
const requires = <T extends ZodRawShape>(schema: z.ZodObject<T>) => {
  const s = schema.strict()
  return (d: any) => {
    const result = s.safeParse(d)
    if (!result.success) {
      throw new ValidationError((result as any).error.issues)
    }
    return result.data
  }
}

// decorator, powered by zod. validates + parses and assigns parsed data to body
function ParseBody(schema: any) {
  function parseBody(req: Request, res: Response, next: NextFunction) {
    const { success, data, error } = schema.safeParse(req.body)
    if (!success) {
      throw new ValidationError(error.issues)
    } else {
      req.body = data
    }

    next()
  }

  parseBody.params = { schema: { zod: schema } }

  return Middleware(parseBody)
}

// decorator, powered by ajv, only validates.
function ValidateBody(schema: any, customValidations = {}) {
  const ajv = new Ajv()
  L.toPairs(customValidations).forEach(([validationName, validationSchema]) => {
    ajv.addKeyword(validationName, validationSchema as any)
  })
  function validateBody(req: Request, res: Response, next: NextFunction) {
    const validator = ajv.compile(schema) // pay compilation because we dont want to share instance
    if (!validator(req.body)) {
      throw new ValidationError(validator.errors)
    }

    next()
  }

  return Middleware(validateBody)
}

export { ValidateBody, ParseBody, permits, requires }
