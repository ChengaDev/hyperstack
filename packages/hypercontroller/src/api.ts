import {
  ClassWrapper,
  ClassErrorMiddleware as OVClassErrorMiddleware,
  ClassMiddleware as OVClassMiddleware,
  Controller as OVController,
  ErrorMiddleware as OVErrorMiddleware,
  Middleware as OVMiddleware,
} from '@hyperstackjs/hypernight'
import asyncWrap from 'async-express-mw'
import L from 'lodash'
import {
  HttpResponseBadRequest,
  HttpResponseNotFound,
  HttpResponseOK,
  HttpResponseUnauthorized,
} from './http/responses'

import responseHandler from './http/response-handler'
import { createServer, listRoutes, listen, load, printRoutes } from './server'

export const ok = (data: any) => new HttpResponseOK(data)
export const unauthorized = (error: any) =>
  new HttpResponseUnauthorized({
    error,
  })
export const err = (error: any) =>
  new HttpResponseBadRequest({
    error,
  })
export const notfound = (error: any) =>
  new HttpResponseNotFound({
    error,
  })
// safe async wrap middleware decorators
const Middleware = (mware: any) =>
  OVMiddleware(L.castArray(mware).map((m) => asyncWrap(m)))

const ClassMiddleware = (mware: any) =>
  OVClassMiddleware(L.castArray(mware).map((m) => asyncWrap(m)))

const ErrorMiddleware = (mware: any) =>
  OVErrorMiddleware(L.castArray(mware).map((m) => asyncWrap(m)))

const ClassErrorMiddleware = (mware: any) =>
  OVClassErrorMiddleware(L.castArray(mware).map((m) => asyncWrap(m)))
// </end safe async decorators>

// decorator
const Controller = (path: string) => (target: any) => {
  const k = OVController(path)(target)
  ClassWrapper(responseHandler)(target)
  return k
}

export {
  load,
  listen,
  listRoutes,
  printRoutes,
  createServer,
  Middleware,
  ClassMiddleware,
  ErrorMiddleware,
  ClassErrorMiddleware,
  Controller,
}

// api decorators
export { Delete, Get, Patch, Post, Put } from '@hyperstackjs/hypernight'
