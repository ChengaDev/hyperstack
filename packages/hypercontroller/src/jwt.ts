import { Bearer } from 'permit'
import { sign, verify } from 'jsonwebtoken'
import type { NextFunction, Request, Response } from './types'
import { ClassMiddleware, Middleware } from './api'
import { HttpResponseUnauthorized } from './http/responses'
const forceHS512 = (jwtOptions: any) =>
  jwtOptions
    ? { ...jwtOptions, algorithm: 'HS512' } // force hard crypto
    : { algorithm: 'HS512' }

const forceHS512Verify = (jwtOptions: any) =>
  jwtOptions
    ? { ...jwtOptions, algorithms: ['HS512'] } // force hard crypto
    : { algorithms: ['HS512'] }

// decorator creator
const createAuth = (verify: any, { loader, bearer, authCookieName }: any) => {
  const permit = new Bearer(
    bearer || {
      query: 'access_token',
    }
  )

  return async (req: Request) => {
    const token = authCookieName
      ? req.cookies[authCookieName]
      : permit.check(req)
    if (!token) {
      throw new HttpResponseUnauthorized('no token')
    }

    let payload: any
    try {
      payload = await verify(token)
    } catch (error) {
      throw new HttpResponseUnauthorized('bad token')
    }

    const detail = await loader(payload)
    if (!detail || !detail.user) {
      throw new HttpResponseUnauthorized('no such user')
    }

    return detail
  }
}

const buildAuth = (verify: any, MWare: any, opts: any) => {
  const auth = createAuth(verify, opts)

  return MWare(async (req: Request, res: Response, next: NextFunction) => {
    const { user, ...rest } = await auth(req)
    req.user = user
    req.state = rest
    next()
  })
}

const authWithJWT = ({
  secret,
  jwtOptions,
  loader,
  bearer,
  authCookieName,
}: {
  secret: string
  loader: any
  jwtOptions?: any
  bearer?: any
  authCookieName?: string | null
}) => {
  if (!secret) {
    throw new Error('no JWT secret set')
  }
  const resolvedOptions = jwtOptions || {
    expiresIn: '14d',
  }

  const signJWT = (payload: any) =>
    sign(payload, secret, forceHS512(resolvedOptions))

  const verifyJWT = (token: string) =>
    new Promise((resolve, reject) => {
      verify(token, secret, forceHS512Verify(resolvedOptions), (err, value) => {
        if (err) {
          reject(err)
        } else {
          resolve(value)
        }
      })
    })

  return {
    MustAuthWithJWT: buildAuth(verifyJWT, ClassMiddleware, {
      loader,
      bearer,
      authCookieName,
    }),
    MustAuthRouteWithJWT: buildAuth(verifyJWT, Middleware, {
      loader,
      bearer,
      authCookieName,
    }),
    signJWT,
    verifyJWT,
  }
}

export { authWithJWT }
