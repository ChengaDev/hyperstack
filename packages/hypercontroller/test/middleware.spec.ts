import request from 'supertest'

import type { Request, Response, ServerOptions } from '../src/index'
import { Controller, Get, HttpResponseOK, createServer } from '../src/index'
import { expectWithSnapshot, logger } from './utils'

@Controller('api')
class Api {
  @Get('foobar')
  async foobar(_req: Request, _res: Response) {
    return new HttpResponseOK({ its: 'alright' })
  }
}

const createMiddlewareServer = (opts: Partial<ServerOptions>) =>
  createServer({
    controllers: [Api],
    opts: { logging: { logger }, sendValidationErrors: true, ...opts },
  })
const matchUuid =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/

describe('middleware/set-request-ids', () => {
  it('none', async () => {
    const app = createMiddlewareServer({ requestId: { mode: 'none' } })
    await expectWithSnapshot(
      200,
      request(app)
        .get('/api/foobar')
        .set({ 'x-request-id': '00000000-bf22-4b1d-be68-b3a4156635c2' }),
      (res: any) => res.headers['x-request-id'] === undefined
    )
    await expectWithSnapshot(
      200,
      request(app).get('/api/foobar'),
      (res: any) => res.headers['x-request-id'] === undefined
    )
  })
  it('passthrough', async () => {
    const app = createMiddlewareServer({ requestId: { mode: 'passthrough' } })
    await expectWithSnapshot(
      200,
      request(app)
        .get('/api/foobar')
        .set({ 'x-request-id': '00000000-bf22-4b1d-be68-b3a4156635c2' }),
      (res: any) => res.headers['x-request-id'].match(/^00000000-.*/)
    )
    await expectWithSnapshot(
      200,
      request(app).get('/api/foobar'),
      (res: any) => res.headers['x-request-id'] === undefined
    )
  })
  it('passthrough-or-generate', async () => {
    const app = createMiddlewareServer({
      requestId: { mode: 'passthrough-or-generate' },
    })
    await expectWithSnapshot(
      200,
      request(app).get('/api/foobar'),
      (res: any) =>
        !res.headers['x-request-id'].match(/^00000000-.*/) &&
        res.headers['x-request-id'].match(matchUuid)
    )
    await expectWithSnapshot(
      200,
      request(app)
        .get('/api/foobar')
        .set({ 'x-request-id': '00000000-bf22-4b1d-be68-b3a4156635c2' }),
      (res: any) => res.headers['x-request-id'].match(/^00000000-.*/)
    )
  })
  it('generate', async () => {
    const app = createMiddlewareServer({ requestId: { mode: 'generate' } })
    await expectWithSnapshot(
      200,
      request(app)
        .get('/api/foobar')
        .set({ 'x-request-id': '00000000-bf22-4b1d-be68-b3a4156635c2' }),
      (res: any) =>
        !res.headers['x-request-id'].match(/^00000000-.*/) &&
        res.headers['x-request-id'].match(matchUuid)
    )
  })
  it('default', async () => {
    const app = createMiddlewareServer({})
    // passthrough - or generate
    await expectWithSnapshot(
      200,
      request(app).get('/api/foobar'),
      (res: any) =>
        !res.headers['x-request-id'].match(/^00000000-.*/) &&
        res.headers['x-request-id'].match(matchUuid)
    )
    await expectWithSnapshot(
      200,
      request(app)
        .get('/api/foobar')
        .set({ 'x-request-id': '00000000-bf22-4b1d-be68-b3a4156635c2' }),
      (res: any) => res.headers['x-request-id'].match(/^00000000-.*/)
    )
  })
})
