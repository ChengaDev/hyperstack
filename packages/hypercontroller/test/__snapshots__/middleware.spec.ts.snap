// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware/set-request-ids default 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids default 2`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids generate 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids none 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids none 2`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids passthrough 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids passthrough 2`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids passthrough-or-generate 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`middleware/set-request-ids passthrough-or-generate 2`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;
